### PR TITLE
Migration from V2 to V3: Integrate and upgrade V2's LiquidityProtection contract + add migration e2e tests

### PR DESCRIPTION
**Pull request by lbeder**

### CODE REVIEW

File: components/LegacyContracts.ts [hunk 1]
```diff
@@ -42,50 +51,74 @@ import {
...
 
 /* eslint-enable camelcase */
 
+export const Registry = {
+    BANCOR_NETWORK: formatBytes32String('BancorNetwork'),
+    NETWORK_SETTINGS: formatBytes32String('NetworkSettings'),
+    CONVERTER_FACTORY: formatBytes32String('ConverterFactory'),
+    CONVERTER_REGISTRY: formatBytes32String('BancorConverterRegistry'),
+    CONVERTER_REGISTRY_DATA: formatBytes32String('BancorConverterRegistryData'),
+    LIQUIDITY_PROTECTION: formatBytes32String('LIQUIDITY_PROTECTION')
```
[the remark this answers is not in the document]
Comment (lbeder) [reply components/LegacyContracts.ts:94]: Yep

File: deploy/tests/network.ts [hunk 2]
```diff
@@ -199,202 +224,333 @@ import { getNamedAccounts } from 'hardhat';
...
+                            n: prevLiquidity.bntTradingLiquidity,
+                            d: prevLiquidity.baseTokenTradingLiquidity
+                        }).to.be.almostEqual(
+                            {
+                                n: liquidity.bntTradingLiquidity,
+                                d: liquidity.baseTokenTradingLiquidity
+                            },
+                            {
+                                maxAbsoluteError: new Decimal(0),
+                                maxRelativeError: new Decimal('0000000000000000000001')
```
[the remark this answers is not in the document]
Comment (lbeder) [reply deploy/tests/network.ts:266]: Interesting. It seems we have this in other places as well.

File: deploy/tests/network.ts [hunk 3]
```diff
@@ -199,202 +224,333 @@ import { getNamedAccounts } from 'hardhat';
...
+                    const prevBNBNTAmount = await getBalance(bnBNT, bntWhale);
+                    const prevVBNTTokenAmount = await getBalance(vbnt, bntWhale);
+                    const prevTotalSupply = await bnt.totalSupply();
+
+                    await bnt.connect(bntWhale).approve(network.address, bntAmount);
+                    await network.connect(bntWhale).deposit(bnt.address, bntAmount);
+
+                    const receivedBNBNTAmount = (await getBalance(bnBNT, bntWhale)).sub(prevBNBNTAmount);
+
+                    expect(receivedBNBNTAmount).be.gt(0);
```
[the remark this answers is not in the document]
Comment (lbeder) [reply deploy/tests/network.ts:287]: 👍

File: test/pools/PoolCollectionWithdrawal.ts [hunk 4]
```diff
@@ -134,11 +134,11 @@ describe('PoolCollectionWithdrawal', () => {
...
         test('PoolCollectionWithdrawalCoverage8', {
-            p: { maxAbsoluteError: new Decimal(0), maxRelativeError: new Decimal('0') },
-            q: { maxAbsoluteError: new Decimal(0), maxRelativeError: new Decimal('0') },
-            r: { maxAbsoluteError: new Decimal(0), maxRelativeError: new Decimal('0') },
+            p: { maxRelativeError: new Decimal('0') },
+            q: { maxRelativeError: new Decimal('0') },
+            r: { maxRelativeError: new Decimal('0') },
             s: { maxAbsoluteError: new Decimal(1), maxRelativeError: new Decimal('0'), relation: Relation.LesserOrEqual }, // prettier-ignore
-            t: { maxAbsoluteError: new Decimal(0), maxRelativeError: new Decimal('0') },
+            t: { maxRelativeError: new Decimal('0') },
```
Comment (barakman) [inline test/pools/PoolCollectionWithdrawal.ts:141]: Well, in this case, I actually left them here on purpose, in order to keep a full depiction of the maximum error in each case. If you insist on removing the `maxAbsoluteError` default ones, then note that we can remove the `maxRelativeError` default ones as well.

Comment (lbeder) [reply test/pools/PoolCollectionWithdrawal.ts:141]: Oh, yeah. Let's remove them as well

File: test/rewards/StandardRewards.ts [hunk 5]
```diff
@@ -2428,8 +2437,7 @@ describe('StandardRewards', () => {
                             }
 
                             expect(totalClaimed).to.be.almostEqual(expectedTotalClaimedReward, {
-                                maxAbsoluteError: new Decimal(0),
-                                maxRelativeError: new Decimal('0000000000000000000001')
+                                maxRelativeError: new Decimal('0.00000000000000001')
```
Comment (barakman) [inline test/rewards/StandardRewards.ts:2440]: Verified all 5 occurrences

**Review by barakman — APPROVED** [review]

**Review by yudilevi — APPROVED** [review]

**Pull request merged**